In [1]:
!pip install scipy

  Using cached scipy-1.9.3-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (33.8 MB)


In [2]:
import tensorflow as tf
import tensorflow.keras as keras
import scipy
from tensorflow.keras.preprocessing.image import ImageDataGenerator

2022-11-21 07:54:06.987760: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.


In [3]:
model = keras.models.Sequential([
    keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(150, 150, 3)),
    keras.layers.MaxPooling2D((2, 2)),
    keras.layers.Flatten(),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(optimizer=keras.optimizers.SGD(learning_rate=0.002, momentum=0.8), loss='binary_crossentropy', metrics=['accuracy'])

2022-11-21 07:54:34.255099: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-21 07:54:34.262390: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-21 07:54:34.263038: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-21 07:54:34.264182: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the approp

Question 1

Since we have a binary classification problem, what is the best loss function for us?

    -> binary crossentropy
    focal loss
    mean squared error
    categorical crossentropy

Note: since we specify an activation for the output layer, we don't need to set from_logits=True

In [5]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 74, 74, 32)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 175232)            0         
                                                                 
 dense (Dense)               (None, 64)                11214912  
                                                                 
 dense_1 (Dense)             (None, 1)                 65        
                                                                 
Total params: 11,215,873
Trainable params: 11,215,873
Non-trainable params: 0
____________________________________________

Question 2

What's the total number of parameters of the model? You can use the summary method for that.

    9215873
    -> 11215873
    14215873
    19215873

In [4]:
train_generator = ImageDataGenerator(rescale=1./255)

In [5]:
train_ds = train_generator.flow_from_directory(
    './train',
    target_size=(150, 150),
    batch_size=20,
    shuffle=True, 
    class_mode='binary'
)

Found 1594 images belonging to 2 classes.


In [24]:
train_ds.class_mode

'binary'

In [6]:
test_generator = ImageDataGenerator(rescale=1./255)
test_ds = test_generator.flow_from_directory(
    './test',
    target_size=(150, 150),
    batch_size=20,
    shuffle=False, 
    class_mode='binary'
)

Found 394 images belonging to 2 classes.


In [12]:
test_ds.class_mode

'binary'

In [7]:
history = model.fit(
    train_ds,
    epochs=10,
    validation_data=test_ds
)

Epoch 1/10


2022-11-21 07:54:54.442963: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8100
2022-11-21 07:54:55.097652: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-21 07:54:55.098611: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-21 07:54:55.098648: W tensorflow/stream_executor/gpu/asm_compiler.cc:80] Couldn't get ptxas version string: INTERNAL: Couldn't invoke ptxas --version
2022-11-21 07:54:55.099697: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-21 07:54:55.099777: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] INTERNAL: Failed to launch ptxas
Relying on driver to perform ptx compilation. 
Modify $PATH to customize ptxas location.
This message will be only logged once.


80/80 [==============================] - 12s 126ms/step - loss: 0.6798 - accuracy: 0.5728 - val_loss: 0.6860 - val_accuracy: 0.5000
Epoch 2/10
80/80 [==============================] - 9s 108ms/step - loss: 0.6045 - accuracy: 0.6945 - val_loss: 0.5864 - val_accuracy: 0.6447
Epoch 3/10
80/80 [==============================] - 9s 107ms/step - loss: 0.5209 - accuracy: 0.7578 - val_loss: 0.5151 - val_accuracy: 0.7335
Epoch 4/10
80/80 [==============================] - 9s 107ms/step - loss: 0.4270 - accuracy: 0.8231 - val_loss: 0.3961 - val_accuracy: 0.8376
Epoch 5/10
80/80 [==============================] - 9s 107ms/step - loss: 0.3821 - accuracy: 0.8369 - val_loss: 0.3941 - val_accuracy: 0.8299
Epoch 6/10
80/80 [==============================] - 9s 106ms/step - loss: 0.3465 - accuracy: 0.8563 - val_loss: 0.3528 - val_accuracy: 0.8426
Epoch 7/10
80/80 [==============================] - 8s 106ms/step - loss: 0.3079 - accuracy: 0.8733 - val_loss: 0.3419 - val_accuracy: 0.8452
Epoch 8/10
80/80

Question 3

What is the median of training accuracy for all the epochs for this model?

    0.40
    0.60
    -> 0.90
    0.20


In [8]:
import numpy as np
np.median(history.history['accuracy'])

0.846612274646759

Question 4

What is the standard deviation of training loss for all the epochs for this model?

    -> 0.11
    0.66
    0.99
    0.33

In [9]:
np.std(history.history["loss"])

0.14498461389631426

Data Augmentation

For the next two questions, we'll generate more data using data augmentations.

Add the following augmentations to your training data generator:

    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest'


In [11]:
train_generator = ImageDataGenerator(
    rescale=1./255,
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest'
)
train_ds = train_generator.flow_from_directory(
    './train',
    target_size=(150, 150),
    batch_size=20,
    shuffle=True, 
    class_mode='binary'
)
test_generator = ImageDataGenerator(rescale=1./255)
test_ds = test_generator.flow_from_directory(
    './test',
    target_size=(150, 150),
    batch_size=20,
    shuffle=False, 
    class_mode='binary'
)

Found 1594 images belonging to 2 classes.
Found 394 images belonging to 2 classes.


Question 5

Let's train our model for 10 more epochs using the same code as previously. Make sure you don't re-create the model - we want to continue training the model we already started training.

What is the mean of test loss for all the epochs for the model trained with augmentations?

    0.15
    0.77
    -> 0.37
    0.97


In [12]:
history = model.fit(
    train_ds,
    epochs=10,
    validation_data=test_ds
)

Epoch 1/10
80/80 [==============================] - 15s 193ms/step - loss: 0.4852 - accuracy: 0.7779 - val_loss: 0.3149 - val_accuracy: 0.8731
Epoch 2/10
80/80 [==============================] - 15s 189ms/step - loss: 0.4245 - accuracy: 0.8162 - val_loss: 0.4044 - val_accuracy: 0.8147
Epoch 3/10
80/80 [==============================] - 15s 191ms/step - loss: 0.4210 - accuracy: 0.8087 - val_loss: 0.4635 - val_accuracy: 0.7919
Epoch 4/10
80/80 [==============================] - 15s 190ms/step - loss: 0.4109 - accuracy: 0.8099 - val_loss: 0.3808 - val_accuracy: 0.8147
Epoch 5/10
80/80 [==============================] - 15s 190ms/step - loss: 0.4279 - accuracy: 0.7861 - val_loss: 0.3178 - val_accuracy: 0.8553
Epoch 6/10
80/80 [==============================] - 15s 190ms/step - loss: 0.4017 - accuracy: 0.8193 - val_loss: 0.2542 - val_accuracy: 0.9086
Epoch 7/10
80/80 [==============================] - 15s 190ms/step - loss: 0.3940 - accuracy: 0.8218 - val_loss: 0.3790 - val_accuracy: 0.8223

In [13]:
np.mean(history.history["val_loss"])

0.375788676738739

In [14]:
np.mean(history.history["val_accuracy"][6:])

0.8236040472984314

Question 6

What's the average of test accuracy for the last 5 epochs (from 6 to 10) for the model trained with augmentations?

    -> 0.84
    0.54
    0.44
    0.24
